### PR TITLE
spring runtime assembly cleaning

### PR DESCRIPTION
**Pull request by ghost**
This mechanical diff factors multiple occurrences of the stack gymkhana done in non-leaf functions, and standardizes on the `ENTER_FUNCTION` and `LEAVE_FUNCTION` found on amd64.

This helps reading the code a bit, and also comparing platforms against each other.

No testsuite regression observed on the four affected platforms.

Not sure this is worth a `Changes` entry as this is quite mechanical and almost doesn't change anything (the astute reader will notice that this fixes a missing stack adjustment in the raising exception path of `caml_call_realloc_stack` on s390x, which was harmless since an exception is raised anyway.)

### CODE REVIEW

**Comment by shindere** [comment]
Not able to review but on the topic of the Changes entry it's also a
good way to credit both authors and reviewers of PRs and to convey an
idea of what is worked on so I would be of the opinion to add an entry
if that's okay with you.

File: runtime/arm64.S [hunk 1]
```diff
@@ -185,6 +185,21 @@ G(name):
...
+        stp     x29, x30, [sp, -16]!
+        CFI_ADJUST(16)
+        add     x29, sp, #0
+.endm
+
+.macro LEAVE_FUNCTION
+        ldp     x29, x30, [sp], 16
+        CFI_ADJUST(-16)
+.endm
+
```
Comment (tmcgilchrist) [inline runtime/arm64.S:171]: On the ordering of CFI directives, should they be before or after the instruction they relate to? When I compile C code to get the assembly, I get the instruction first and then the CFI directive. For compiled OCaml code and runtime assembly there is a mix of before and after. Does it matter? For example on entry to a main function:

```
_main:                                  ; @main
	.cfi_startproc
; %bb.0:
	sub	sp, sp, #64
	.cfi_def_cfa_offset 64
	stp	x29, x30, [sp, #48]             ; 16-byte Folded Spill
	add	x29, sp, #48
	.cfi_def_cfa w29, 16
	.cfi_offset w30, -8
	.cfi_offset w29, -16
   ...
```

**Review by tmcgilchrist — COMMENTED** [review]
I am more familiar with power and arm64 assembly, which look correct to me, and the other backends look like a straight forward replacement. General question about the correct use of CFI directives, this PR follows the existing patterns so not specific to this change.

File: runtime/arm64.S (hunk 1 above)
Comment (ghost) [reply runtime/arm64.S:171]: My understanding is that CFI annotations must appear after the instruction which causes the changes they refer to.

**Comment by tmcgilchrist** [comment]
Looks good to me. 👍🏻

**Review by tmcgilchrist — APPROVED** [review]

**Review by gasche — APPROVED** [review]
Approved on behalf of @tmcgilchrist. Thanks everyone!

**Comment by gasche** [comment]
Hmm, it looks like the CI is failing on `macos-arm64`. Many lines that look like this:

```

<instantiation>:1:14: error: expected ')'
CFI_OFFSET(29, -16)
             ^
runtime/arm64.S:438:9: note: while in macro instantiation
        ENTER_FUNCTION
        ^
<instantiation>:1:11: error: invalid operand
CFI_OFFSET(29, -16)
          ^
runtime/arm64.S:438:9: note: while in macro instantiation
        ENTER_FUNCTION
        ^
```

**Comment by ghost** [comment]
> Hmm, it looks like the CI is failing on `macos-arm64`. Many lines that look like this:
> 
> ```
> 
> <instantiation>:1:14: error: expected ')'
> CFI_OFFSET(29, -16)
>              ^
> runtime/arm64.S:438:9: note: while in macro instantiation
>         ENTER_FUNCTION
>         ^
> <instantiation>:1:11: error: invalid operand
> CFI_OFFSET(29, -16)
>           ^
> runtime/arm64.S:438:9: note: while in macro instantiation
>         ENTER_FUNCTION
>         ^
> ```

This should be fixed now. I moved the inclusion of `../runtime/caml/asm.h` above the declaration of the macros which use `CFI_OFFSET`.

**Comment by ghost** [comment]
> This should be fixed now. I moved the inclusion of `../runtime/caml/asm.h` above the declaration of the macros which use `CFI_OFFSET`.

...and `riscv64.S` needed the same change.

The other backends do not because they use `cpp` macros rather than `as` macros.

**Pull request merged**